lib/Dialect/FIRRTL/Transforms/LowerClasses.cpp

```diff
@@ -36,6 +36,38 @@ using namespace circt::om;
 
 namespace {
 
+static bool shouldCreateClassImpl(igraph::InstanceGraphNode *node) {
+  auto moduleLike = node->getModule<FModuleLike>();
+  if (!moduleLike)
+    return false;
+
+  if (isa<firrtl::ClassLike>(moduleLike.getOperation()))
+    return true;
+
+  // Always create a class for public modules.
+  if (moduleLike.isPublic())
+    return true;
+
+  // Create a class for modules with property ports.
+  bool hasClassPorts = llvm::any_of(moduleLike.getPorts(), [](PortInfo port) {
```

Comment:
Just thinking out loud -- doing this in parallel when it walked the module body made sense, but maybe less so to walk the ports + instance nodes for each?

Something like:
"has property ports" can be computed once per module (and looked up instance-side, instead of re-computed, so tracked in a (temporary?) data structure), and "shouldCreateClassImpl" can then be determined in a single (serial) bottom-up walk of instance graph, computing "hasPropertyPorts" for module definitions along the way.

WDYT?

Reply:
I think that's great idea. We have to fix InstanceGraph for bottom-up walk so let me first merge the PR. I'll follow-up to use a bottom-up walk.